{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rotary-finger",
   "metadata": {},
   "outputs": [],
   "source": [
    "drad_fasta_file = \"/mnt/data/lizards/darevskia_genomes_analysis/drad/drad_ragoo_softmasked.fasta\"\n",
    "duni_fasta_file = \"/mnt/data/lizards/darevskia_genomes_analysis/duni/duni_ragoo_softmasked.fasta\"\n",
    "dval_fasta_file = \"/mnt/data/lizards/darevskia_genomes_analysis/dval/dval_assembly2020A_softmasked.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "impossible-converter",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from dna_features_viewer import GraphicFeature, GraphicRecord, CircularGraphicRecord, BiopythonTranslator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "outside-blanket",
   "metadata": {},
   "outputs": [],
   "source": [
    "## делаем словарь из фасты\n",
    "def read_fasta_file(fasta_file):\n",
    "    fasta = {}\n",
    "    header = None\n",
    "    with open(fasta_file) as fh:\n",
    "        for i, line in enumerate(fh):\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if header:\n",
    "                    fasta[header] = \"\".join(seq)\n",
    "                header = line\n",
    "                seq = []\n",
    "            else:\n",
    "                seq.append(line)\n",
    "        if header:\n",
    "            fasta[header] = \"\".join(seq)\n",
    "    return fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "superior-value",
   "metadata": {},
   "outputs": [],
   "source": [
    "drad_reference = read_fasta_file(drad_fasta_file)\n",
    "duni_reference = read_fasta_file(duni_fasta_file)\n",
    "dval_reference = read_fasta_file(dval_fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "owned-arrest",
   "metadata": {},
   "outputs": [],
   "source": [
    "## D.raddei\n",
    "\n",
    "drad_SMC2 = [\n",
    "    GraphicFeature(start=90647, end=108033, strand=+1, color=\"#00FF7F\", label=\"SMC2\"),\n",
    "    GraphicFeature(start=31392, end=48531, strand=-1, color=\"#A9A9A9\", label=\"PTGR1\"),\n",
    "    GraphicFeature(start=66522, end=67031, strand=+1, color=\"#A9A9A9\", label=\"TUSC1\"),\n",
    "    GraphicFeature(start=54, end=29015, strand=+1, color=\"#A9A9A9\", label=\"SVEP1\"),\n",
    "    GraphicFeature(start=155694, end=185154, strand=+1, color=\"#A9A9A9\", label=\"MOB3B \")\n",
    "]\n",
    "record=GraphicRecord(sequence_length = len(drad_reference[\">79385_RaGOO\"]), features = drad_SMC2)\n",
    "start, stop = 0, 200000\n",
    "cropped_record = record.crop((start, stop))\n",
    "cropped_record.plot(figure_width = 7)\n",
    "\n",
    "drad_SMC4 = [\n",
    "    GraphicFeature(start=6299317, end=6332267, strand=-1, color=\"#FF4500\", label=\"SMC4\"),\n",
    "    GraphicFeature(start=6245019, end=6273712, strand=+1, color=\"#A9A9A9\", label=\"KPNA4\"),\n",
    "    GraphicFeature(start=6150080, end=6204204, strand=+1, color=\"#A9A9A9\", label=\"PPM1L\"),\n",
    "    GraphicFeature(start=6339212, end=6396913, strand=+1, color=\"#A9A9A9\", label=\"IFT80 \"),\n",
    "    GraphicFeature(start=6459008, end=6485189, strand=+1, color=\"#A9A9A9\", label=\"CARD8\"),\n",
    "    GraphicFeature(start=6411818, end=6412807, strand=-1, color=\"#A9A9A9\", label=\"C3orf80\")\n",
    "]\n",
    "record=GraphicRecord(sequence_length = len(drad_reference[\">77_RaGOO\"]), features = drad_SMC4)\n",
    "start2, stop2 = 6000000, 6500000\n",
    "cropped_record = record.crop((start2, stop2))\n",
    "cropped_record.plot(figure_width = 7)\n",
    "\n",
    "\n",
    "drad_SMC3 = [\n",
    "    GraphicFeature(start=6385604, end=6415166, strand=-1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "    GraphicFeature(start=6215383, end=6217334, strand=+1, color=\"#A9A9A9\", label=\"BBIP1\"),\n",
    "    GraphicFeature(start=6249679, end=6355975, strand=-1, color=\"#A9A9A9\", label=\"RBM20\"),\n",
    "    GraphicFeature(start=6556814, end=6566579, strand=-1, color=\"#A9A9A9\", label=\"MXI1\"),\n",
    "    GraphicFeature(start=6426616, end=6435262, strand=-1, color=\"#A9A9A9\", label=\"DUSP5\"),\n",
    "     GraphicFeature(start=6159195, end=6181868, strand=-1, color=\"#A9A9A9\", label=\"SHOC2\")\n",
    "]\n",
    "record=GraphicRecord(sequence_length = len(drad_reference[\">71_RaGOO\"]), features = drad_SMC3)\n",
    "start2, stop2 = 6120000, 6600000\n",
    "cropped_record = record.crop((start2, stop2))\n",
    "cropped_record.plot(figure_width = 7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dominant-effect",
   "metadata": {},
   "outputs": [],
   "source": [
    "## выделяем нужный кусок аннотации\n",
    "ann = \"/mnt/data/lizards/sochkalova/annotation/braker/augustus.hints.gtf\"\n",
    "with open(ann) as cc:\n",
    "    for line1 in cc:\n",
    "        line1 = line1.split()\n",
    "        if \"#\" not in line1:\n",
    "            if line1[0] == \"72\" and line1[2] == \"transcript\" and int(line1[3]) in range(4400000, 4525000):\n",
    "                print(line1)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "headed-partition",
   "metadata": {},
   "outputs": [],
   "source": [
    "## D. unsisexualis \n",
    "duni_SMC2_1 = [\n",
    "    GraphicFeature(start=26719067, end=26726062, strand=+1, color=\"#00FF7F\", label=\"SMC2\"),\n",
    "    GraphicFeature(start=26751725, end=26755346, strand=+1, color=\"#A9A9A9\", label=\"manba\"),\n",
    "    GraphicFeature(start=26551920, end=26590515, strand=+1, color=\"#A9A9A9\", label=\"PCNXL3\"),\n",
    "    GraphicFeature(start=26646627, end=26648173, strand=+1, color=\"#A9A9A9\", label=\"CINP\"),\n",
    "    GraphicFeature(start=26688975, end=26712595, strand=+1, color=\"#A9A9A9\", label=\"VPS13D\"),\n",
    "    GraphicFeature(start=26756696, end=26764429, strand=-1, color=\"#A9A9A9\", label=\"PRICKLE3\"), \n",
    "    GraphicFeature(start=26598767, end=26613783, strand=-1, color=\"#A9A9A9\", label=\"BUD31\"), \n",
    "    GraphicFeature(start=26654871, end=26659096, strand=-1, color=\"#A9A9A9\", label=\"MSH3\"), \n",
    "]\n",
    "record=GraphicRecord(sequence_length = len(duni_reference[\">Chr0_RaGOO\"]), features = duni_SMC2_1)\n",
    "start, stop = 26550000, 26800000\n",
    "cropped_record = record.crop((start, stop))\n",
    "cropped_record.plot(figure_width = 8)\n",
    "\n",
    "duni_SMC2_2 = [\n",
    "    GraphicFeature(start=92947, end=114427, strand=-1, color=\"#00FF7F\", label=\"SMC2\"),\n",
    "    GraphicFeature(start=30401, end=45672, strand=-1, color=\"#A9A9A9\", label=\"PTGR1 \"),\n",
    "    GraphicFeature(start=30401, end=47987, strand=-1, color=\"#A9A9A9\", label=\"PTGR1 \"),\n",
    "    GraphicFeature(start=71870, end=72388, strand=+1, color=\"#A9A9A9\", label=\"TUSC1\"),\n",
    "    GraphicFeature(start=115, end=29030, strand=+1, color=\"#A9A9A9\", label=\"SVEP1\"),\n",
    "    GraphicFeature(start=115, end=29030, strand=+1, color=\"#A9A9A9\", label=\"SVEP1\"),\n",
    "    GraphicFeature(start=115, end=29030, strand=+1, color=\"#A9A9A9\", label=\"SVEP1\")\n",
    "]\n",
    "record=GraphicRecord(sequence_length = len(duni_reference[\">79385_RaGOO\"]), features = duni_SMC2_2)\n",
    "start2, stop2 = 0, 150000 \n",
    "cropped_record = record.crop((start2, stop2))\n",
    "cropped_record.plot(figure_width = 8)\n",
    "\n",
    "duni_SMC4 = [\n",
    "    GraphicFeature(start=7664654, end=7712599, strand=-1, color=\"#FF4500\", label=\"SMC4\"),\n",
    "    GraphicFeature(start=7664654, end=7712599, strand=-1, color=\"#FF4500\", label=\"SMC4\"),\n",
    "    GraphicFeature(start=7664654, end=7712599, strand=-1, color=\"#FF4500\", label=\"SMC4\"),\n",
    "    GraphicFeature(start=7664654, end=7686650, strand=-1, color=\"#FF4500\", label=\"SMC4\"),\n",
    "    GraphicFeature(start=7661334, end=7712599, strand=-1, color=\"#FF4500\", label=\"SMC4\"),\n",
    "    GraphicFeature(start=7656760, end=7657572, strand=+1, color=\"#A9A9A9\", label=\"TRIM59\"),\n",
    "    GraphicFeature(start=7722176, end=7790308, strand=+1, color=\"#A9A9A9\", label=\"IFT80\"),\n",
    "    GraphicFeature(start=7722176, end=7790308, strand=+1, color=\"#A9A9A9\", label=\"IFT80\"),\n",
    "]\n",
    "\n",
    "record=GraphicRecord(sequence_length = len(duni_reference[\">77_RaGOO\"]), features = duni_SMC4)\n",
    "zoom_start, zoom_end = 7650000, 7800000  # coordinates of the \"detail\"\n",
    "cropped_record = record.crop((zoom_start, zoom_end))\n",
    "cropped_record.plot(figure_width = 8)\n",
    "\n",
    "duni_SMC3_1 = [\n",
    "    GraphicFeature(start=7578948, end=7599472, strand=-1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "    GraphicFeature(start=7576352, end=7599472, strand=-1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "    GraphicFeature(start=7578948, end=7599472, strand=-1, color=\"#FFD700\", label=\"SMC3\"),  \n",
    "    GraphicFeature(start=7623146, end=7632082, strand=-1, color=\"#A9A9A9\", label=\"DUSP5\")\n",
    "]\n",
    "record=GraphicRecord(sequence_length = len(duni_reference[\">71_RaGOO\"]), features = duni_SMC3_1)\n",
    "cropped_record = record.crop((7500000, 7650000))\n",
    "cropped_record.plot(figure_width = 8)\n",
    "\n",
    "duni_SMC3_2 = [\n",
    "    GraphicFeature(start=38, end=8851, strand=+1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "    GraphicFeature(start=38, end=8851, strand=+1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "]\n",
    "record=GraphicRecord(sequence_length = len(duni_reference[\">76206_RaGOO\"]), features = duni_SMC3_2)\n",
    "record.plot(figure_width = 8)\n",
    "\n",
    "\"\"\"\"\"\n",
    "duni_SMC3_3 = [\n",
    "    GraphicFeature(start=38, end=8851, strand=+1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "    GraphicFeature(start=38, end=8851, strand=+1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "]\n",
    "record=GraphicRecord(sequence_length = len(duni_reference[\">76206_RaGOO\"]), features = duni_SMC3_2)\n",
    "record.plot(figure_width = 15)\n",
    "\"\"\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "primary-redhead",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "loose-fossil",
   "metadata": {},
   "outputs": [],
   "source": [
    "### D.VALENTINI\n",
    "dval_SMC2 = [\n",
    "    GraphicFeature(start=2541609, end=2553857, strand=+1, color=\"#00FF7F\", label=\"SMC2\"),\n",
    "    GraphicFeature(start=2648954, end=2651886, strand=+1, color=\"#A9A9A9\", label=\"LRRC19\"),\n",
    "    GraphicFeature(start=2653515, end=2686800, strand=+1, color=\"#A9A9A9\", label=\"IFT74 \"),\n",
    "    GraphicFeature(start=2594129, end=2600765, strand=+1, color=\"#A9A9A9\", label=\"REXO1\"),\n",
    "    GraphicFeature(start=2698295, end=2709390, strand=+1, color=\"#A9A9A9\", label=\"PLAA\"),\n",
    "    GraphicFeature(start=2600783, end=2638662, strand=+1, color=\"#A9A9A9\", label=\"TEK\")\n",
    "]\n",
    "record = GraphicRecord(sequence_length = len(dval_reference[\">139\"]), features = dval_SMC2)\n",
    "start, stop = 2500000, 2700000\n",
    "cropped_record = record.crop((start, stop))\n",
    "cropped_record.plot(figure_width = 8)\n",
    "\n",
    "dval_SMC4 = [\n",
    "    GraphicFeature(start=1024837, end=1062658, strand=-1, color=\"#FF4500\", label=\"SMC4\"),\n",
    "    GraphicFeature(start=1099309, end=1128307, strand=-1, color=\"#A9A9A9\", label=\"C3orf80\"),\n",
    "    GraphicFeature(start=1099309, end=1099309, strand=-1, color=\"#A9A9A9\", label=\"C3orf80\"),\n",
    "    GraphicFeature(start=1067255, end=1086832, strand=-1, color=\"#A9A9A9\", label=\"IFT80\"),\n",
    "    GraphicFeature(start=1168320, end=1169092, strand=-1, color=\"#A9A9A9\", label=\"CARD8\")\n",
    "]\n",
    "record = GraphicRecord(sequence_length = len(dval_reference[\">107\"]), features = dval_SMC4)\n",
    "cropped_record = record.crop((800000, 1200000))\n",
    "cropped_record.plot(figure_width = 8)\n",
    "\n",
    "dval_SMC3_1 = [\n",
    "    GraphicFeature(start=1697, end=4873, strand=-1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "    GraphicFeature(start=4979, end=6142, strand=+1, color=\"#A9A9A9\", label=\"SMNDC1\"),\n",
    "    \n",
    "]\n",
    "record = GraphicRecord(sequence_length = len(dval_reference[\">76721\"]), features = dval_SMC3_1)\n",
    "#cropped_record = record.crop((800000, 1200000))\n",
    "#cropped_record.plot(figure_width = 15)\n",
    "record.plot(figure_width = 8)\n",
    "\n",
    "dval_SMC3_2 = [\n",
    "    GraphicFeature(start=4502733, end=4506126, strand=+1, color=\"#FFD700\", label=\"SMC3\"),\n",
    "    GraphicFeature(start=4467677, end=4472167, strand=+1, color=\"#A9A9A9\", label=\"KRT4\"),\n",
    "    GraphicFeature(start=4483711, end=4489980, strand=+1, color=\"#A9A9A9\", label=\"NR5A1\")\n",
    "]\n",
    "record = GraphicRecord(sequence_length = len(dval_reference[\">72\"]), features = dval_SMC3_2)\n",
    "cropped_record = record.crop((4400000, 4525000))\n",
    "cropped_record.plot(figure_width = 8)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "increasing-philosophy",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "instructional-registrar",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
